src/inmanta/ast/statements/generator.py

```diff
@@ -268,7 +267,159 @@ def nested_blocks(self) -> Iterator["BasicBlock"]:
+
+    def requires(self) -> list[str]:
+        # exclude loop var, unless it shadows an occurrence in iterable
+        return list(set(self.iterable.requires()) - {self.loop_var} + set(self.value_expression.requires()))
+
+    def requires_emit(
+        self, resolver: Resolver, queue: QueueScheduler, *, lhs: Optional[ResultCollector[object]] = None
+    ) -> dict[object, VariableABC]:
+        """
+        Sets up gradual execution of the list comprehension. Additionally sets up a resumer for when the iterable is complete.
+        Returns as requires the gradual helper's result variable, which will be frozen by the resumer.
+
+        Flow: requires_emit -> gradual execution -> resume -> execute
+        """
+        base_requires: dict[object, VariableABC] = super().requires_emit(resolver, queue)
+
+        # set up gradual execution
+        gradual_helper: ListComprehensionGradual = ListComprehensionGradual(
+            statement=self,
+            resolver=resolver,
+            queue=queue,
+            lhs=lhs,
+        )
+        iterable_requires: dict[object, VariableABC] = self.iterable.requires_emit_gradual(
+            resolver, queue, resultcollector=gradual_helper
+        )
+
+        # non-gradual mode / finishing up
+        # pass helper to the resumer via the requires object
+        wrapped_helper: VariableABC = WrappedValueVariable(gradual_helper)
+        requires: dict[object, VariableABC] = base_requires | iterable_requires | {self: wrapped_helper}
+        RawUnit(queue, resolver, requires, resumer=self)
+
+        # wait for resumer to populate result
+        # TODO: consider whether base_requires and/or iterable_requires are still important here
+        return {self: gradual_helper.result}
+
+    def requires_emit_gradual(
+        self, resolver: Resolver, queue: QueueScheduler, resultcollector: ResultCollector[object]
+    ) -> dict[object, VariableABC]:
+        return self.requires_emit(resolver, queue, lhs=resultcollector)
+
+    def resume(self, requires: dict[object, VariableABC], resolver: Resolver, queue: QueueScheduler) -> None:
+        """
+        Resume when the iterable is fully ready for execution. Populates the helper's result variable and signals that no more
+        values will be sent.
+        """
+        # fetch helper, passed via the requires object
+        gradual_helper: object = requires[self].get_value()
+        assert isinstance(gradual_helper, ListComprehensionGradual)
+
+        iterable: object = self.iterable.execute({k: v.get_value() for k, v in requires.items()}, resolver, queue)
+        if isinstance(iterable, Unknown):
+            return Unknown(self)
+        if not isinstance(iterable, list):
+            # TODO: allow dict?
+            raise TypingException(
+                self, f"A list comprehension can only be applied to lists and relations, got {type(iterable)}"
+            )
+        if gradual_helper.nb_results_received == 0:
```

Comment:
I don't know if a mixed mode can occur, but the fact that we don't de-duplicate makes that rather important.

Reply:
Note to self: double check that all statements that implement `requires_emit_gradual` have a consistent `execute` and document that as a requirement.

Reply:
I think we concluded that we can't currently think of a mixed scenario, so it doesn't make it worse.